csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -743,15 +764,23 @@ class AllocationInfoMap : private kir::IrVisitor {
 
   void handle(kir::ForLoop* for_loop) final {
     auto loop_info = scope_map_.getLoopScopeInfo(for_loop);
-    current_stack_.push_back(loop_info);
+    if (!for_loop->iter_domain()->isParallelized()) {
+      // Parallelized loops do not result in for loops in the CUDA kernel, so
+      // they should not affect liveness analysis. This means that
+      // current_stack_ will differ from kir::IrVisitor::for_loops_, which will
+      // actually hold all ForLoops regardless of parallelization.
+      current_stack_.push_back(loop_info);
```

Comment:
I noticed this when testing with matmuls; see `FusionAmpereMatmulSmemEpilogue_CUDA` for example. In that case the kernel starts with
```
T9_s[ ... ] ca_pos( 2 ) = ALLOCATE(buffer=T9_s[ ... ] ca_pos( 2 ), mem_type=shared, size=8192, zero_init=false)
T8_s[ ... ] ca_pos( 2 ) = ALLOCATE(buffer=T8_s[ ... ] ca_pos( 2 ), mem_type=shared, size=4096, zero_init=false)
T7_s[ ... ] ca_pos( 2 ) produce_pos( 2 ) = ALLOCATE(buffer=T7_s[ ... ] ca_pos( 2 ) produce_pos( 2 ), mem_type=shared, size=8192, zero_init=false)
FOR blockIdx.x in iblockIdx.x71{( ceilDiv(T0.logical_size[0], 64) )}:
  FOR blockIdx.y in iblockIdx.y73{( ceilDiv(T1.logical_size[1], 128) )}:
  ...
    FOR threadIdx.z in ithreadIdx.z84{( ceilDiv(( ( ceilDiv(64, 4) ) * 4 ), 32) )}:
      FOR threadIdx.y in ithreadIdx.y86{( ceilDiv(( ( ( ceilDiv(( ceilDiv(128, 8) ), 4) ) * 4 ) * 8 ), 32) )}:
        FOR i806 in iS88{( ceilDiv(32, 16) )}:  
```
Here we see that only the `i806` loop will be present in the actual CUDA code. The parallelized loops cover the entire kernel, so without this change, the outer live interval of every tensor is just the range of the outermost `BIDx` loop. With this change, we properly compute the live intervals relative to loops that should occur in the kernel.

This change is related to this PR but not strictly necessary for it. We could always remove it for now and reintroduce it when we make the change to the matmul scheduler.

Reply:
Regarding the example at https://github.com/NVIDIA/Fuser/blob/b92150300fdf6b294121757134ac795aed14b66c/csrc/device_lower/pass/alias_memory.cpp#L926-L938
What happens if I have:
```C++
  //!  For ...
  //!    For ...
  //!      Allocate    <---- reference arg
  //!      For ..
  //!          For ...
  //!      For blockIdx.x in blockDim.x <---- Will this function returns `ScopeInfo` for this loop?
  //!          For ...
  //!             expr  <---- current expr (implied in current_stack_ and current_pos_ )
```

Reply:
Since the `For blockIdx.x in blockDim.x` loop is parallelized, it will not appear in the kernel. So in that case this should look like
```
  //!  For ...
  //!    For ...
  //!      Allocate    <---- reference arg
  //!      For ..
  //!          For ...
  //!      // For blockIdx.x in blockDim.x <---- This loop does not appear in the CUDA code, so it is ignored
  //!      For ...  <---- This function returns `ScopeInfo` for this loop
  //!         expr  <---- current expr (implied in current_stack_ and current_pos_ )
```

Reply:
@jacobhinkle Could you remind me what the problem is with:

> so without this change, the outer live interval of every tensor is just the range of the outermost BIDx loop.

Reply:
The problem is that in that case we actually could do "outer sharing", since in the cuda kernel the live intervals do not overlap. However, because the trivial BIDx loop surrounds the whole kernel, without this change the outer live interval of any allocation is simply the span of that BIDx loop.

Reply:
It seems this could affect many more cases (positively). Can you split out this change from this PR and also see if how the aliasing would change with the benchmarks?

Reply:
Also, there's a couple of places we do `current_stack_.back()` and I wonder they are all safe. If all loops are parallelized, wouldn't the stack just be empty?

Reply:
Yes I will go ahead and split this off into another PR.

Reply:
Moved to #766.